### PR TITLE
Join drep endpoint

### PR DESCRIPTION
**Pull request by paweljakubas**
- [x] Swagger updated
- [x] Api scaffolding, Api types, Api errors
- [x] Adding Http.Shelley.Server joinDRep
- [x] Adding IO Deleg functions
- [x] Adding Delegation's guard
- [x] Checking that error is emitted when used in Babbage
- [x] Checking we can vote, re-vote and emit error when voting the same in the absence of pool delegation in Conway
- [x] Checking we can vote, re-vote and emit error when voting the same in the presence of pool delegation in Conway

### Issue Number
adp-3383

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet/IO/Delegation.hs [hunk 1]
```diff
@@ -528,3 +494,109 @@ quitStakePool ctx walletId passphrase = do
...
+
+    either (throwIO . ExceptionVoting) pure
+        (WD.guardOnlyVoting era $ toDrepEnriched votingRequest)
+    pure vAction
+  where
+    toDrepEnriched NotVotedYet = Nothing
+    toDrepEnriched VotedSameAsBefore = Just (True, drep)
+    toDrepEnriched VotedDifferently = Just (False, drep)
+
+    netLayer = ctx ^. networkLayer
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/IO/Delegation.hs]: This function is okay, but harder to understand than necessary — in particular, it's not clear what exactly `toDrepEnrichted` is supposed to be doing. Could you remove this function (`handleVoteRequest`) and structure the function `joinDRep` similar to the function `quitStakePool` instead?

In other words, I would like you to structure the code in the following way:

```hs
joinDRep — IO, collects relevant parameter values, calls

    joinDRepVotingAction — IO, provides additional parameter values (if necessary), calls

        WD.joinDRepVotingAction — *pure*, computes a result based on all parameter values.
```

The idea is that the functions `joinDRep` and `joinDRepVotingAction` are responsible for providing sufficient parameter values to compute the result, but the actual computation of the result (= the `Either ErrCannotJoin Tx.VotingAction` is contained within the pure function. In other words, the concern of figuring out what the vote action should be — if any — should be entirely confined to the definition of the pure function `WD.joinDRepVotingAction`.

In contrast, at the moment, this function uses `toDRepEnriched` to package some intermediate computation and hands it off to the function `WD.guardOnlyVoting ` in a different module, but the meaning of the intermediate data format `Maybe (Bool, DRep)` is not apparent from the type signatures and function names. The logic becomes much clear if:

* Use pure functions to compute the result.
* Keep the computation of the result to a single module

Could you change this?

**Review by HeinrichApfelmus — COMMENTED** [review]
Thank you! 😊 

The functionality seems to work, but I request a different separation of concerns for the modules `Cardano.Wallet.Delegation` and `Cardano.Wallet.IO.Delegation`.

File: lib/wallet/src/Cardano/Wallet/IO/Delegation.hs (hunk 1 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/IO/Delegation.hs]: sounds reasonable! done

File: lib/wallet/src/Cardano/Wallet/Delegation.hs [hunk 2]
```diff
@@ -208,3 +217,42 @@ guardVoting
...
+
+    isSameNext (W.WalletDelegationNext _ deleg) = isDRepSame deleg
+
+    sameWalletDelegation (W.WalletDelegation current coming) =
+        isDRepSame current || any isSameNext coming
+
+    (vAction, votingRequest) =
+        if stakeKeyIsRegistered
+        then (Tx.Vote action, (sameWalletDelegation dlg, action))
+        else (Tx.VoteRegisteringKey action, (sameWalletDelegation dlg,action))
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Delegation.hs]:
```suggestion
joinDRepVotingAction
    :: Write.IsRecentEra era
    => Write.RecentEra era
    -> DRep
    -> W.WalletDelegation
    -> Bool
    -> Either ErrCannotVote Tx.VotingAction
joinDRepVotingAction era targetDRep dlg stakeKeyIsRegistered = do
    guardEraIsConway era
    when (sameWalletDelegation dlg) $
        Left $ ErrAlreadyVoted $ targetDRep
    pure votingAction
  where
    isDRepSame (W.Voting drep) = drep == targetDRep
    isDRepSame (W.DelegatingVoting _ drep) = drep == targetDRep
    isDRepSame _ = False

    isSameNext (W.WalletDelegationNext _ deleg) = isDRepSame deleg

    sameWalletDelegation (W.WalletDelegation current coming) =
        isDRepSame current || any isSameNext coming

    guardEraIsConway Write.RecentEraBabbage = Left WrongEra
    guardEraIsConway Write.RecentEraConway = Right ()

    votingAction =
        if stakeKeyIsRegistered
        then Tx.Vote targetDRep
        else Tx.VoteRegisteringKey targetDRep
```

… and after separating the concerns into a single, we can now merge the definitions of the two functions and get rid of the `(Bool,DRep)` intermediate type while also making the logic more clear. 🤓

**Review by HeinrichApfelmus — APPROVED** [review]
Thank you! 😊 

With the improved separation of concerns for `joinDRep`, we can now perform a further simplification in `Cardano.Wallet.Delegation` — good to merge after that.

(In principle, the same considerations apply to the `Cardano.Wallet.IO.Delegation.joinStakePool` function, but let's leave this for a future pull request when we need to touch this area of the code again.)

File: lib/wallet/src/Cardano/Wallet/Delegation.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/Delegation.hs]: good hint. Done!

**Pull request merged**